### PR TITLE
 Extend needless_pass_by_value to methods 

### PR DESCRIPTION
**Pull request by sinkuu**
I don't remember why I didn't include them initially.

### CODE REVIEW

File: tests/ui/needless_pass_by_value.rs [hunk 1]
```diff
@@ -79,4 +79,27 @@ fn issue_2114(s: String, t: String, u: Vec<i32>, v: Vec<i32>) {
...
+    fn bar(
+        _t: T, // Ok, since `&T: Serialize` too
+    ) {
+    }
+
+    fn baz(
+        &self,
+        _u: U,
+    ) {
+    }
```
Comment (oli-obk) [inline tests/ui/needless_pass_by_value.rs]: can you add a test for another argument that is of type `Self`? Just that we document this behaviour?

File: clippy_lints/src/utils/sugg.rs [hunk 2]
```diff
@@ -162,10 +162,10 @@ impl<'a> Sugg<'a> {
 
     /// Convenience method to create the `<lhs>..<rhs>` or `<lhs>...<rhs>`
     /// suggestion.
-    pub fn range(self, end: Self, limit: ast::RangeLimits) -> Sugg<'static> {
+    pub fn range<E: Borrow<Self>>(self, end: E, limit: ast::RangeLimits) -> Sugg<'static> {
```
Comment (oli-obk) [inline clippy_lints/src/utils/sugg.rs]: Any reason not to simply use `&Self`?

File: clippy_lints/src/needless_pass_by_value.rs [hunk 3]
```diff
@@ -22,13 +23,20 @@ use std::borrow::Cow;
 /// sometimes avoid
 /// unnecessary allocations.
 ///
-/// **Known problems:** Hopefully none.
+/// **Known problems:**
+/// * This lint suggests taking an argument by reference,
+/// however sometimes it is better to let users decide the argument type
+/// (by using `Borrow` trait, for example), depending on how the function is used.
```
Comment (oli-obk) [inline clippy_lints/src/needless_pass_by_value.rs]: If we suggest this action, we should probably also generate a second structured suggestion that adds the generic argument

Comment (sinkuu) [reply clippy_lints/src/needless_pass_by_value.rs]: Hmm I just wanted to note it somewhere as an alternative to taking by reference.

File: clippy_lints/src/utils/sugg.rs (hunk 2 above)
Comment (sinkuu) [reply clippy_lints/src/utils/sugg.rs]: I thought it could be better for these combinator-like things, and was too lazy to look for callsites...

Comment (oli-obk) [reply clippy_lints/src/utils/sugg.rs]: > too lazy to look for callsites...

I can totally get behind that :D

**Pull request merged**